core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1503,7 +1503,10 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     def sendResponseCallback(producerId: Long, result: TransactionResult)(responseStatus: Map[TopicPartition, PartitionResponse]): Unit = {
       trace(s"End transaction marker append for producer id $producerId completed with status: $responseStatus")
-      val partitionErrors = responseStatus.mapValues(_.error).asJava
+      val partitionErrors = new util.HashMap[TopicPartition, Errors]()
+      responseStatus.foreach { case (topicPartition, partitionResponse) =>
+        partitionErrors.put(topicPartition, partitionResponse.error)
+      }
       val previous = errors.putIfAbsent(producerId, partitionErrors)
       if (previous != null)
         previous.putAll(partitionErrors)
```

Comment:
I've been thinking about this. The only way we could hit the UnsupportedOperationException is if the WriteTxnMarkerRequest has multiple entries for the same producerId. I assumed this was impossible initially and just did this:
```
errors.put(producerId, responseStatus.mapValues(_.error).asJava)
```
Later this was changed to the current logic. Now, I think it's unsafe to make the assumption generally that there is only one entry in the request per producerId, so fixing this makes sense rather than reverting. However, it seems to indicate a bug in the grouping logic in `TransactionMarkerChannelManager`. 

cc @guozhangwang